### PR TITLE
11futures

### PR DESCRIPTION
**Pull request by markuslamm**
first version, please review @lauraluiz @schleichardt

### CODE REVIEW

File: app/controllers/OrderController.java [hunk 1]
```diff
@@ -3,58 +3,67 @@
...
-            final int selectedFrequency = cartService.getFrequency(currentCart.getId());
-            if (selectedFrequency > 0) {
-                final ProductVariant selectedVariant = currentCart.getLineItems().get(0).getVariant();
-                final OrderPageData orderPageData = new OrderPageData(selectedVariant, selectedFrequency, currentCart);
-                return ok(order.render(orderPageData));
-            } else {
-                flash("error", "Missing frequency of delivery. Please try selecting it again.");
+    public F.Promise<Result> show() {
+        LOG.debug("Display Order details page");
+        final F.Promise<Cart> currentCartPromise = cartService._getOrCreateCart(session());
```
Comment (schleichardt) [inline app/controllers/OrderController.java]: I would not keep the method names with the underscore

File: app/controllers/PactasWebhookController.java [hunk 2]
```diff
@@ -43,27 +47,20 @@ public Result createOrderFromSubscription() {
...
-                final Order order = orderService.createOrder(cart);
-                Logger.debug("Order created: {}", order);
-                return ok();
-            } catch (PactasException e) {
-                Logger.error(e.getMessage(), e);
-                return internalServerError();
-            } catch (Exception e) {
-                Logger.error(e.getMessage(), e);
-            }
+            final PactasContract contract = pactas.fetchContract(contractId.get()).get(2000, TimeUnit.MILLISECONDS);
```
Comment (schleichardt) [inline app/controllers/PactasWebhookController.java]: this part is not async

File: app/inject/DonutShopModule.java [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+import play.Configuration;
+import play.Environment;
+import play.Logger;
+import services.*;
+
+import javax.inject.Singleton;
+
+import static java.util.Objects.requireNonNull;
+
+public class DonutShopModule extends AbstractModule {
```
Comment (schleichardt) [inline app/inject/DonutShopModule.java]: You don't close the sphere client when the application are supposed to be shutdown.

File: app/services/OrderService.java [hunk 4]
```diff
@@ -15,4 +16,6 @@
      * @return the newly created {@code Order}, must not be null
      */
     Order createOrder(Cart cart);
```
Comment (schleichardt) [inline app/services/OrderService.java]: I would delete the blocking methods if there is already an async method.

File: app/controllers/OrderController.java [hunk 5]
```diff
@@ -3,58 +3,67 @@
...
-        try {
-            final Cart currentCart = cartService.getOrCreateCart(session());
-            final Cart clearedCart = cartService.clearCart(currentCart);
-            return ok(success.render());
-        } catch (Exception e) {
-            Logger.error(e.getMessage(), e);
-            return internalServerError();
-        }
+    //TODO check is doing nothing than clear the cart?!
+    public F.Promise<Result> submit() {
```
Comment (lauraluiz) [inline app/controllers/OrderController.java]: You can't do anything else, the order is created when the customer pays the subscription each month, not when creating the subscription. But I think the customer (representing the subscription) could be saved and then link the order with the customer, that would be a nice feature. But it's just a wild thought.

File: app/controllers/OrderController.java [hunk 6]
```diff
@@ -3,58 +3,67 @@
...
+        LOG.debug("Submitting Order details page");
+        final F.Promise<Cart> currentCartPromise = cartService.getOrCreateCart(session());
+        return currentCartPromise.map(currentCart -> ok(success.render()));
```
Comment (lauraluiz) [inline app/controllers/OrderController.java]: But certainly you don't have to wait for the cart, just call `success.render()` directly.

Comment (lauraluiz) [inline app/controllers/OrderController.java]: In here you will eventually have to clear the session, once https://github.com/commercetools/sphere-donut/issues/29 is implemented. Also maybe the customer according to https://github.com/commercetools/sphere-donut/issues/30

File: app/services/ProductServiceImpl.java [hunk 7]
```diff
@@ -1,43 +1,40 @@
...
-        return Optional.ofNullable(product.getVariant(variantId));
+    @Inject
+    public ProductServiceImpl(final ApplicationLifecycle applicationLifecycle,
+                              final PlayJavaSphereClient playJavaSphereClient) {
+        super(applicationLifecycle, playJavaSphereClient);
+        this.cachedProduct = loadProduct();
     }
 
-    private Optional<ProductProjection> loadProduct() {
+    private F.Promise<Optional<ProductProjection>> loadProduct() {
```
Comment (lauraluiz) [inline app/services/ProductServiceImpl.java]: Now that we use DI, wouldn't it be better to obtain this product in the `DonutShopModule` and get rid of this class?

File: it/services/CartServiceIntegrationTest.java [hunk 8]
```diff
@@ -0,0 +1,99 @@
...
+        Http.Context.current.set(context);
+
+        productController = fakeApplication.injector().instanceOf(ProductController.class);
+        cartService = fakeApplication.injector().instanceOf(CartService.class);
+        productService = fakeApplication.injector().instanceOf(ProductService.class);
+    }
+
+    @Test
+    public void testGetOrCreateCart() {
+        final Cart cart = cartService.getOrCreateCart(productController.session()).get(2000);
```
Comment (lauraluiz) [inline it/services/CartServiceIntegrationTest.java]: I would extract all `2000` into a constant `ALLOWED_TIMEOUT` or similar.

**Comment by schleichardt** [comment]
@markuslamm can you fix the pull request, please

**Pull request merged**